cmd/osbuild-jobsite-builder/main.go

```diff
@@ -0,0 +1,406 @@
+	agent.StateLock.Lock()
+	defer agent.StateLock.Unlock()
+
+	if state <= agent.State {
+		agent.State = StateError
+	} else {
+		agent.State = state
+	}
+
+	agent.StateChannel <- agent.State
+}
+
+func (agent *Agent) GetState() State {
+	agent.StateLock.Lock()
+	defer agent.StateLock.Unlock()
+
+	return agent.State
+}
+
+func (agent *Agent) GuardState(stateWanted State) {
+	if stateCurrent := agent.GetState(); stateWanted != stateCurrent {
+		logrus.Fatalf("Agent.GuardState: Requested guard for %d but we're in %d. Exit.", stateWanted, stateCurrent)
+	}
+}
+
+func (agent *Agent) HandleClaim(w http.ResponseWriter, r *http.Request) {
+	agent.GuardState(StateClaim)
+
+	if r.Method != "POST" {
+		logrus.WithFields(
+			logrus.Fields{"method": r.Method},
+		).Fatal("Agent.HandleClaim: unexpected request method")
+	}
+
+	fmt.Fprintf(w, "%s", "done")
+
+	logrus.Info("Agent.HandleClaim: Done.")
+
+	agent.SetState(StateProvision)
+}
+
+func (agent *Agent) HandleProvision(w http.ResponseWriter, r *http.Request) {
+	agent.GuardState(StateProvision)
+
+	if r.Method != "PUT" {
+		logrus.WithFields(
+			logrus.Fields{"method": r.Method},
+		).Fatal("Agent.HandleProvision: Unexpected request method.")
+	}
+
+	logrus.WithFields(logrus.Fields{"argBuildPath": argBuildPath}).Debug("Agent.HandleProvision: Opening manifest.json.")
+
+	dst, err := os.OpenFile(
+		path.Join(argBuildPath, "manifest.json"),
+		os.O_WRONLY|os.O_CREATE|os.O_EXCL,
+		0400,
+	)
+	defer dst.Close()
+
+	if err != nil {
+		logrus.Fatal("Agent.HandleProvision: Failed to open manifest.json.")
+	}
+
+	logrus.Debug("Agent.HandleProvision: Writing manifest.json.")
+
+	_, err = io.Copy(dst, r.Body)
+
+	if err != nil {
+		logrus.Fatal("Agent.HandleProvision: Failed to write manifest.json.")
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Write([]byte(`done`))
+
+	logrus.Info("Agent.HandleProvision: Done.")
+
+	agent.SetState(StatePopulate)
+}
+
+func (agent *Agent) HandlePopulate(w http.ResponseWriter, r *http.Request) {
+	agent.GuardState(StatePopulate)
+
+	if r.Method != "POST" {
+		logrus.WithFields(
+			logrus.Fields{"method": r.Method},
+		).Fatal("Agent.HandlePopulate: unexpected request method")
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Write([]byte(`done`))
+
+	logrus.Info("Agent.HandlePopulate: Done.")
+
+	agent.SetState(StateBuild)
+}
+
+func (agent *Agent) HandleBuild(w http.ResponseWriter, r *http.Request) {
+	agent.GuardState(StateBuild)
+
+	if r.Method != "POST" {
+		logrus.WithFields(
+			logrus.Fields{"method": r.Method},
+		).Fatal("Agent.HandleBuild: Unexpected request method.")
+	}
+
+	var buildRequest BuildRequest
+
+	var err error
+
+	if err = json.NewDecoder(r.Body).Decode(&buildRequest); err != nil {
+		logrus.Fatal("HandleBuild: Failed to decode body.")
+	}
+
+	if Build != nil {
+		logrus.Fatal("HandleBuild: what?")
+	}
+
+	args := []string{
+		"--store", path.Join(argBuildPath, "store"),
+		"--cache-max-size", "unlimited",
+		"--checkpoint", "*",
```

Comment:
We can decide on a reasonable `cache-max-size`, @croissanne how much disk are we planning to give these you think? We could even go with no caching initially since we aren't re-using jobsites.

Reply:
The current workers have a 50GB disk, so probably around the same?

Reply:
I think we should be a bit more defensive here. Checkpoints and caching is not useful at all with ephemeral workers. So I suggest turning it off for now just to keep things minimal. We can change it when we have a proper caching story for jobsites.

Reply:
Ok.